lms/djangoapps/courseware/local_decisions/permissions.rst

```diff
@@ -0,0 +1,119 @@
+Plan of Action
+==============
+
+#. Subclass ``Predicate`` to allow for non-booleans to be returned during
+   evaluation.
+#. Convert built-in predicates to non-boolean predicates in edx-platform.
+#. For each caller of ``has_access``:
+
+   #. Convert the caller to use ``user.has_perm`` instead.
+   #. Implement the new permission created in 1. by referencing to the previous
+      ``has_access`` call.
+
+#. Refactor contents of ``has_access`` out into their own predicates that can
+   be used to implement specific permissions.
+#. For each place that checks if a user is enrolled in a specific track:
```

Comment:
Is there a pattern for setting up the permissions, separate from the predicates? [OEP-9](https://github.com/edx/open-edx-proposals/blob/b97380497b8d63bf7e686c451bba26f7136fdf16/oeps/oep-0009-bp-permissions.rst#rule-based-authorization) suggests that each app should create predicates in rules.py without registering permissions, but it does not recommend where in edx-platform to add the actual permissions. I can imagine that some of the predicates will be complex.

Reply:
My expectation is that as we're converting things from `has_access` or track membership testing, we'll create names for the various permissions being checked, and then bind those new permissions to particular predicates in the appropriate `rules.py` files.

Reply:
Because the permissions are just string names, I think we're going to have to use best-judgement for those at the moment. Maybe once we get more in the codebase, we can start rationalizing them a bit, but I think it's premature to try and constrain them just yet. Although, dot-separate strings seems like maybe a reasonable standard to start with.

Reply:
> bind those new permissions to particular predicates in the appropriate rules.py files.

For instance, edx-proctoring may have a `rules.py` file that defines some predicates. The proctoring code will check against some permission name -- `can_view_proctoring` (or something). But that permission should be defined in edx-platform, since it may incorporate other predicates, such as an enrollment mode check, waffle flag, etc., which don't make sense to exist in edx-proctoring because edx-proctoring doesn't depend on edx-platform.

It would be nice if we had a story for how pluggable django apps can announce which permissions they depend on.

Reply:
Hm... Yeah, I'll think on that. The rules library has some discussion on extending permissions definitions with additional predicate checks, but I'll need to read it again to see how it should work w/ the pluggable django apps.

Reply:
On further thought... I feel like for now, anyway, that's really just a documentation task. There's no real automated way for a pluggable django app to request that the surrounding system generate a permission for it. The app could use `add_perm` to add a default permission implementation, and catch the error if that permission already had a rule created for it.